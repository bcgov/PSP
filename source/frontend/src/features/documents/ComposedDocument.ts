import {
  Api_Document,
  Api_DocumentMetadataUpdate,
  Api_DocumentType,
  Api_DocumentUpdateRequest,
  Api_DocumentUploadRequest,
} from 'models/api/Document';
import {
  Api_Storage_DocumentMetadata,
  Api_Storage_DocumentTypeMetadataType,
} from 'models/api/DocumentStorage';

export interface ComposedDocument {
  mayanMetadata?: Api_Storage_DocumentMetadata[];
  pimsDocument?: Api_Document;
  mayanFileId?: number;
}

export class DocumentUploadFormData {
  public documentTypeId: string;
  public documentStatusCode: string;
  public documentMetadata: Record<string, string>;

  public constructor(
    initialStatus: string,
    documentType: string,
    metadata: Api_Storage_DocumentTypeMetadataType[],
  ) {
    this.documentStatusCode = initialStatus;
    this.documentTypeId = documentType;
    this.documentMetadata = {};

    metadata.forEach(metaType => {
      this.documentMetadata[metaType.metadata_type?.id?.toString() || '-'] = '';
    });
  }

  public toRequestApi(file: File, documentType: Api_DocumentType): Api_DocumentUploadRequest {
    var metadata: Api_DocumentMetadataUpdate[] = [];
    for (const key in this.documentMetadata) {
      const value = this.documentMetadata[key];
      metadata.push({
        metadataTypeId: Number(key),
        value: value,
      });
    }

    return {
      documentType: documentType,
      documentStatusCode: this.documentStatusCode,
      file: file,
      documentMetadata: metadata,
    };
  }
}

export class DocumentUpdateFormData {
  public documentId: number;
  public mayanDocumentId: number;
  public documentStatusCode: string;
  public documentMetadata: Record<string, string>;

  public static fromApi(
    composedDocument: ComposedDocument,
    metadataTypes: Api_Storage_DocumentTypeMetadataType[],
  ): DocumentUpdateFormData {
    var model = new DocumentUpdateFormData();
    model.documentId = composedDocument.pimsDocument?.id || 0;
    model.mayanDocumentId = composedDocument.pimsDocument?.mayanDocumentId || 0;
    model.documentStatusCode = composedDocument.pimsDocument?.statusTypeCode?.id?.toString() || '';
    model.documentMetadata = {};
    metadataTypes.forEach(metaType => {
      var foundMetadata = composedDocument.mayanMetadata?.find(
        currentMeta => currentMeta.metadata_type.id === metaType.metadata_type?.id,
      );
      model.documentMetadata[metaType.metadata_type?.id?.toString() || '-'] =
        foundMetadata?.value ?? '';
    });
    return model;
  }

  public toRequestApi(): Api_DocumentUpdateRequest {
    var metadata: Api_DocumentMetadataUpdate[] = [];

    for (const key in this.documentMetadata) {
      const value = this.documentMetadata[key];
      const metadataTypeId = Number(key);
      metadata.push({
        value: value,
        metadataTypeId: metadataTypeId,
      });
    }

    return {
      documentId: this.documentId,
      mayanDocumentId: this.mayanDocumentId,
      documentStatusCode: this.documentStatusCode,
      documentMetadata: metadata,
    };
  }

  private constructor() {
    this.documentId = -1;
    this.mayanDocumentId = -1;
    this.documentStatusCode = '';
    this.documentMetadata = {};
  }
}
