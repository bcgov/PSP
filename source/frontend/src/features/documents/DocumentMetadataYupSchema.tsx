import * as Yup from 'yup';

import { ApiGen_Mayan_DocumentTypeMetadataType } from '@/models/api/generated/ApiGen_Mayan_DocumentTypeMetadataType';

export const getDocumentMetadataYupSchema = (
  mayanMetadata: ApiGen_Mayan_DocumentTypeMetadataType[],
) => {
  const metadataSchema: Record<string, Yup.StringSchema> = {};
  for (const data of mayanMetadata) {
    if (data.metadata_type?.id && data.required) {
      metadataSchema[`${data.metadata_type?.id}`] = Yup.string().required(
        `${data.metadata_type?.label} is required`,
      );
    }
  }

  return Yup.object().shape({
    documentMetadata: Yup.object().shape(metadataSchema),
    documentTypeId: Yup.string().required('Document type is required'),
  });
};
