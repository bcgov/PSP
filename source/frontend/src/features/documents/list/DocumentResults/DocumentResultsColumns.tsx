import { StyledRemoveLinkButton } from 'components/common/buttons';
import { Button } from 'components/common/buttons/Button';
import TooltipIcon from 'components/common/TooltipIcon';
import { ColumnWithProps, renderTypeCode } from 'components/Table';
import { Claims } from 'constants/index';
import DownloadDocumentButton from 'features/documents/DownloadDocumentButton';
import useKeycloakWrapper from 'hooks/useKeycloakWrapper';
import { Api_Document, Api_DocumentType } from 'models/api/Document';
import { Col, Row } from 'react-bootstrap';
import { FaEye, FaTrash, FaUserAlt } from 'react-icons/fa';
import { CellProps } from 'react-table';
import styled from 'styled-components';
import { prettyFormatDate } from 'utils';

export interface IDocumentColumnProps {
  onViewDetails: (values: Api_Document) => void;
  onDelete: (values: Api_Document) => void;
}

export const getDocumentColumns = ({
  onViewDetails,
  onDelete,
}: IDocumentColumnProps): ColumnWithProps<Api_Document>[] => {
  return [
    {
      Header: 'Document type',
      accessor: 'documentType',
      align: 'left',
      sortable: true,
      Cell: renderDocumentType,
    },
    {
      Header: 'File name',
      accessor: 'fileName',
      sortable: true,
      Cell: renderFileName(onViewDetails),
    },
    {
      Header: 'Uploaded',
      accessor: 'appCreateTimestamp',
      sortable: true,
      Cell: renderUploaded,
    },
    {
      Header: 'Status',
      accessor: 'statusTypeCode',
      sortable: true,
      Cell: renderTypeCode,
    },
    {
      Header: 'Actions',
      width: '90',
      Cell: renderActions(onViewDetails, onDelete),
    },
  ];
};

function renderDocumentType({ value }: CellProps<Api_Document, Api_DocumentType>) {
  return value?.documentType ?? '';
}

const renderFileName = (onViewDetails: (values: Api_Document) => void) => {
  return function (cell: CellProps<Api_Document, string>) {
    const { hasClaim } = useKeycloakWrapper();
    return (
      <>
        {hasClaim(Claims.DOCUMENT_VIEW) === true ? (
          <Button
            data-testid="document-view-filename-link"
            onClick={() => cell.row.original?.id && onViewDetails(cell.row.original)}
            variant="link"
          >
            {cell.value}
          </Button>
        ) : (
          <span data-testid="document-view-filename-text">{cell.value}</span>
        )}
      </>
    );
  };
};

//export const DateCell = ({ cell: { value } }: CellProps<Date, string>) => prettyFormatDate(value);

function renderUploaded(cell: CellProps<Api_Document, string>) {
  return (
    <Row className="no-gutters">
      <Col>{prettyFormatDate(cell.row.original.appCreateTimestamp)}</Col>
      <Col xs="auto">
        <StyledIcon>
          <TooltipIcon
            toolTipId="initiator-tooltip"
            toolTip={cell.row.original.appCreateUserid}
            customToolTipIcon={<FaUserAlt size={15} />}
          />
        </StyledIcon>
      </Col>
    </Row>
  );
}

const renderActions = (
  onViewDetails: (values: Api_Document) => void,
  onDelete: (values: Api_Document) => void,
) => {
  return function ({ row: { original, index } }: CellProps<Api_Document, string>) {
    const { hasClaim } = useKeycloakWrapper();
    return (
      <StyledIconsRow className="no-gutters">
        {hasClaim(Claims.DOCUMENT_VIEW) && original?.mayanDocumentId !== undefined && (
          <Col>
            <DownloadDocumentButton mayanDocumentId={original?.mayanDocumentId} />
          </Col>
        )}
        {hasClaim(Claims.DOCUMENT_VIEW) && (
          <Col>
            <Button
              data-testid="document-view-button"
              icon={<FaEye size={24} id={`document-view-${index}`} title="document view details" />}
              onClick={() => original?.id && onViewDetails(original)}
            ></Button>
          </Col>
        )}
        {hasClaim(Claims.DOCUMENT_DELETE) && (
          <Col>
            <StyledRemoveLinkButton
              title="document delete"
              data-testid="document-delete-button"
              icon={<FaTrash size={24} id={`document-delete-${index}`} title="document delete" />}
              onClick={() => original?.id && onDelete(original)}
            ></StyledRemoveLinkButton>
          </Col>
        )}
      </StyledIconsRow>
    );
  };
};

const StyledIconsRow = styled(Row)`
  [id^='document-view'] {
    color: ${props => props.theme.css.slideOutBlue};
  }
  [id^='document-delete'] {
    color: ${props => props.theme.css.discardedColor};
    :hover {
      color: ${({ theme }) => theme.css.dangerColor};
    }
  }
  .btn.btn-primary {
    background-color: transparent;
    padding: 0;
    margin-left: 0.5rem;
  }
`;

const StyledIcon = styled.span`
  .tooltip-icon {
    color: ${({ theme }) => theme.css.subtleColor};
  }
`;
