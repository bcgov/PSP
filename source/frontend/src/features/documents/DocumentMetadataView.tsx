import { Input } from 'components/common/form';
import { SectionField } from 'features/mapSideBar/tabs/SectionField';
import { FieldArray, FormikProps } from 'formik';
import { Api_Storage_DocumentTypeMetadataType } from 'models/api/DocumentStorage';

import { StyledNoData } from './commonStyles';
import { DocumentUpdateFormData, DocumentUploadFormData } from './ComposedDocument';

export interface IDocumentMetadataViewProps {
  mayanMetadata: Api_Storage_DocumentTypeMetadataType[];
  formikProps: FormikProps<DocumentUploadFormData> | FormikProps<DocumentUpdateFormData>;
}

export const DocumentMetadataView: React.FunctionComponent<
  React.PropsWithChildren<IDocumentMetadataViewProps>
> = props => {
  return (
    <>
      {props.mayanMetadata.length === 0 && <StyledNoData>No additional data</StyledNoData>}

      <FieldArray
        name="documentMetadata"
        render={() => (
          <>
            {props.mayanMetadata.map(meta => (
              <SectionField
                labelWidth="4"
                key={`document-metadata-${meta.metadata_type?.name}`}
                label={meta.metadata_type?.label || ''}
                required={meta.required === true}
              >
                <Input
                  data-testid={`metadata-input-${meta.metadata_type?.name}` || ''}
                  field={`documentMetadata.${meta.metadata_type?.id}`}
                  required={meta.required === true}
                  value={props.formikProps.values.documentMetadata[meta.metadata_type?.id || '']}
                />
              </SectionField>
            ))}

            <div style={{ color: 'red' }}>
              {Object.values(props.formikProps.errors).length > 0 && (
                <>Mandatory fields are required.</>
              )}
            </div>
          </>
        )}
      />
    </>
  );
};
