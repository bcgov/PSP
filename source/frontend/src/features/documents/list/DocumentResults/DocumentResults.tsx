import { Table } from 'components/Table';
import { TableSort } from 'components/Table/TableSort';
import { DocumentRow } from 'features/documents/ComposedDocument';
import { Api_Document, Api_DocumentRelationship } from 'models/api/Document';
import { useMemo } from 'react';
import { getPage } from 'utils';

import { getDocumentColumns } from './DocumentResultsColumns';

export interface IDocumentResultProps {
  results: DocumentRow[];
  loading?: boolean;
  sort: TableSort<Api_Document>;
  setSort: (value: TableSort<Api_Document>) => void;
  onViewDetails: (values: Api_DocumentRelationship) => void;
  onDelete: (values: Api_DocumentRelationship) => void;
  onPageChange: (props: { pageIndex?: number; pageSize: number }) => void;
  pageProps: { pageIndex?: number; pageSize: number };
}

export const DocumentResults: React.FunctionComponent<
  React.PropsWithChildren<IDocumentResultProps>
> = ({ results, setSort, sort, onViewDetails, onDelete, pageProps, ...rest }) => {
  const columns = useMemo(
    () => getDocumentColumns({ onViewDetails, onDelete }),
    [onViewDetails, onDelete],
  );

  return (
    <Table<DocumentRow>
      name="documentsTable"
      manualSortBy={false}
      manualPagination
      onPageSizeChange={size => {
        rest.onPageChange({ pageSize: size });
      }}
      onRequestData={rest.onPageChange}
      totalItems={results.length}
      columns={columns}
      externalSort={{ sort, setSort }}
      data={getPage(pageProps.pageIndex ?? 0, pageProps.pageSize, results) ?? []}
      noRowsMessage="No matching Documents found"
      pageCount={Math.ceil(results.length / pageProps.pageSize)}
      pageSize={pageProps.pageSize}
      pageIndex={pageProps.pageIndex}
      {...rest}
    ></Table>
  );
};
