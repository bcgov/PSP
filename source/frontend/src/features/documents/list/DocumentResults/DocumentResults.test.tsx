import { useKeycloak } from '@react-keycloak/web';
import noop from 'lodash/noop';

import { Claims } from '@/constants/claims';
import { DocumentRow } from '@/features/documents/ComposedDocument';
import { mockDocumentResponse, mockDocumentsResponse } from '@/mocks/documents.mock';
import { cleanup, mockKeycloak, render, RenderOptions, userEvent } from '@/utils/test-utils';

import { DocumentResults, IDocumentResultProps } from './DocumentResults';
import { get } from 'lodash';
import { ApiGen_CodeTypes_DocumentRelationType } from '@/models/api/generated/ApiGen_CodeTypes_DocumentRelationType';

const setSort = vi.fn();

const onViewDetails = vi.fn();
const onDelete = vi.fn();
const onPreview = vi.fn();
const onViewParent = vi.fn();

// render component under test
const setup = (renderOptions: RenderOptions & Partial<IDocumentResultProps> = { results: [] }) => {
  const { results, ...rest } = renderOptions;
  const utils = render(
    <DocumentResults
      sort={{}}
      results={results ?? []}
      setSort={setSort}
      onViewDetails={onViewDetails}
      onDelete={onDelete}
      onPreview={onPreview}
      onViewParent={onViewParent}
      showParentInformation={false}
    />,
    {
      ...rest,
      useMockAuthentication: true,
    },
  );
  const tableRows = utils.container.querySelectorAll('.table .tbody .tr-wrapper');
  // long css selector to: get the FIRST header or table, then get the SVG up/down sort buttons
  const sortButtons = utils.container.querySelector(
    '.table .thead .tr .sortable-column div',
  ) as HTMLElement;
  return {
    ...utils,
    tableRows,
    sortButtons,
    // Finding elements
    getDocumentFileNameLink: (documentId: number) =>
      utils.container.querySelector(
        `button#document-view-filename-link-${documentId}`,
      ) as HTMLElement,
    getDocumentFileNameText: (documentId: number) =>
      utils.container.querySelector(
        `span#document-view-filename-text-${documentId}`,
      ) as HTMLElement,
    getDocumentProcessingIcon: (rowNumber: number) =>
      utils.container.querySelector(`svg#document-processing-${rowNumber}`) as SVGElement,
    getDocumentProcessingErrorIcon: (rowNumber: number) =>
      utils.container.querySelector(`svg#document-error-${rowNumber}`) as SVGElement,
    getDocumenDeleteIcon: (rowNumber: number) =>
      utils.container.querySelector(`svg#document-delete-${rowNumber}`) as SVGElement,
  };
};

describe('Document Results Table', () => {
  beforeEach(() => {
    mockKeycloak({ claims: [] });
    setSort.mockClear();
  });
  afterEach(() => {
    cleanup();
  });
  afterAll(() => {
    vi.restoreAllMocks();
  });

  it('matches snapshot', async () => {
    const { asFragment } = setup({
      results: mockDocumentsResponse().map(x =>
        DocumentRow.fromApi(x, ApiGen_CodeTypes_DocumentRelationType.Leases),
      ),
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays a message when no matching records found', async () => {
    const { tableRows, findAllByText } = setup({ results: [] });

    expect(tableRows.length).toBe(0);
    const toasts = await findAllByText('No matching Documents found');
    expect(toasts[0]).toBeVisible();
  });

  it('displays document view button', async () => {
    const { getAllByTestId } = setup({
      results: mockDocumentsResponse().map(x =>
        DocumentRow.fromApi(x, ApiGen_CodeTypes_DocumentRelationType.Leases),
      ),
      claims: [Claims.DOCUMENT_VIEW, Claims.DOCUMENT_EDIT],
    });

    const viewButtons = await getAllByTestId('document-view-button');
    expect(viewButtons[0]).toBeVisible();
  });

  it('displays document filename as link', async () => {
    const { getDocumentFileNameLink, getDocumentFileNameText } = setup({
      results: mockDocumentsResponse().map(x =>
        DocumentRow.fromApi(x, ApiGen_CodeTypes_DocumentRelationType.Leases),
      ),
      claims: [Claims.DOCUMENT_VIEW],
    });

    const filenameLink = await getDocumentFileNameLink(20);
    expect(filenameLink).toBeVisible();

    expect(getDocumentFileNameText(20)).toBeNull();
  });

  it('displays document filename as plain text', async () => {
    mockKeycloak({ claims: [] });
    const { getDocumentFileNameLink, getDocumentFileNameText } = setup({
      results: mockDocumentsResponse().map(x =>
        DocumentRow.fromApi(x, ApiGen_CodeTypes_DocumentRelationType.Leases),
      ),
    });

    const filenameText = await getDocumentFileNameText(20);
    expect(filenameText).toBeVisible();

    expect(getDocumentFileNameLink(20)).toBeNull();
  });

  it('displays document processing icon', async () => {
    const { getDocumentProcessingIcon } = setup({
      results: mockDocumentsResponse().map(x =>
        DocumentRow.fromApi(x, ApiGen_CodeTypes_DocumentRelationType.Leases),
      ),
      claims: [Claims.DOCUMENT_VIEW, Claims.DOCUMENT_DELETE],
    });

    expect(getDocumentProcessingIcon(2)).toBeVisible();
  });

  it('displays document PIMS ERROR processing icon', async () => {
    const { getDocumentProcessingErrorIcon } = setup({
      results: mockDocumentsResponse().map(x =>
        DocumentRow.fromApi(x, ApiGen_CodeTypes_DocumentRelationType.Leases),
      ),
      claims: [Claims.DOCUMENT_VIEW, Claims.DOCUMENT_DELETE],
    });

    expect(getDocumentProcessingErrorIcon(3)).toBeVisible();
  });

  it('displays document MAYAN ERROR processing icon', async () => {
    const { getDocumentProcessingErrorIcon } = setup({
      results: mockDocumentsResponse().map(x =>
        DocumentRow.fromApi(x, ApiGen_CodeTypes_DocumentRelationType.Leases),
      ),
      claims: [Claims.DOCUMENT_VIEW, Claims.DOCUMENT_DELETE],
    });

    expect(getDocumentProcessingErrorIcon(4)).toBeVisible();
  });

  it('displays document delete button', async () => {
    const { getAllByTestId } = setup({
      results: mockDocumentsResponse().map(x =>
        DocumentRow.fromApi(x, ApiGen_CodeTypes_DocumentRelationType.Leases),
      ),
      claims: [Claims.DOCUMENT_VIEW, Claims.DOCUMENT_DELETE],
    });

    const deleteButtons = await getAllByTestId('document-delete-button');
    expect(deleteButtons[0]).toBeVisible();
  });

  it('displays default number of entries of 10', async () => {
    const largeDataset = Array.from({ length: 15 }, (id: number) =>
      DocumentRow.fromApi(mockDocumentResponse(id), ApiGen_CodeTypes_DocumentRelationType.Leases),
    );
    const { findByText } = setup({ results: largeDataset, claims: [Claims.DOCUMENT_VIEW] });
    expect(await findByText('1 - 10 of 15')).toBeVisible();
  });

  it('previews a document when text clicked', async () => {
    const { getDocumentFileNameLink } = setup({
      results: mockDocumentsResponse().map(x =>
        DocumentRow.fromApi(x, ApiGen_CodeTypes_DocumentRelationType.Leases),
      ),
      claims: [Claims.DOCUMENT_VIEW],
    });

    const filenameLink = await getDocumentFileNameLink(20);
    userEvent.click(filenameLink);

    expect(onPreview).toHaveBeenCalled();
  });

  it('views a document when eye icon clicked', async () => {
    const { getAllByTestId } = setup({
      results: mockDocumentsResponse().map(x =>
        DocumentRow.fromApi(x, ApiGen_CodeTypes_DocumentRelationType.Leases),
      ),
      claims: [Claims.DOCUMENT_VIEW, Claims.DOCUMENT_EDIT],
    });

    const viewButton = getAllByTestId('document-view-button')[0];
    userEvent.click(viewButton);
    expect(onViewDetails).toHaveBeenCalled();
  });

  it('deletes a document when delete icon clicked', async () => {
    const { getAllByTestId } = setup({
      results: mockDocumentsResponse().map(x =>
        DocumentRow.fromApi(x, ApiGen_CodeTypes_DocumentRelationType.Leases),
      ),
      claims: [Claims.DOCUMENT_VIEW, Claims.DOCUMENT_DELETE],
    });

    const deleteButton = getAllByTestId('document-delete-button')[0];
    userEvent.click(deleteButton);
    expect(onDelete).toHaveBeenCalled();
  });
});
