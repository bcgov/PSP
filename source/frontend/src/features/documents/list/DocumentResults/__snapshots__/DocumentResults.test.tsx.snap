// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document Results Table matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div
    class="container"
  />
  <div
    class="Toastify"
  />
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-left: 0.5rem;
}

.c2 {
  width: 1.6rem;
  height: 1.6rem;
}

.c1 {
  width: 1.6rem;
  height: 1.6rem;
}

.c3 .btn.btn-primary {
  background-color: transparent;
  padding: 0;
  margin-left: 0.5rem;
}

<div
    class="table"
    data-testid="documentsTable"
    role="table"
  >
    <div
      class="thead thead-light"
    >
      <div
        class="tr"
        role="row"
        style="display: flex; flex: 1 0 auto; min-width: 177px;"
      >
        <div
          class="th"
          colspan="1"
          role="columnheader"
          style="box-sizing: border-box; flex: 40 0 auto; min-width: 40px; width: 40px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
        >
          <div
            class="sortable-column"
          >
            Document type
            <div
              class="c0"
              data-testid="sort-column-documentType"
            >
              <svg
                class="c1"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="margin-bottom: -8px;"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M128 320l128-128 128 128z"
                />
              </svg>
              <svg
                class="c2"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M128 192l128 128 128-128z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="th"
          colspan="1"
          role="columnheader"
          style="box-sizing: border-box; flex: 40 0 auto; min-width: 30px; width: 40px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
        >
          <div
            class="sortable-column"
          >
            File name
            <div
              class="c0"
              data-testid="sort-column-fileName"
            >
              <svg
                class="c1"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="margin-bottom: -8px;"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M128 320l128-128 128 128z"
                />
              </svg>
              <svg
                class="c2"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M128 192l128 128 128-128z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="th"
          colspan="1"
          role="columnheader"
          style="box-sizing: border-box; flex: 20 0 auto; min-width: 30px; width: 20px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
        >
          <div
            class="sortable-column"
          >
            Upload date
            <div
              class="c0"
              data-testid="sort-column-appCreateTimestamp"
            >
              <svg
                class="c1"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="margin-bottom: -8px;"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M128 320l128-128 128 128z"
                />
              </svg>
              <svg
                class="c2"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M128 192l128 128 128-128z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="th"
          colspan="1"
          role="columnheader"
          style="box-sizing: border-box; flex: 20 0 auto; min-width: 30px; width: 20px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
        >
          <div
            class="sortable-column"
          >
            Uploaded by
            <div
              class="c0"
              data-testid="sort-column-appCreateUserid"
            >
              <svg
                class="c1"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="margin-bottom: -8px;"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M128 320l128-128 128 128z"
                />
              </svg>
              <svg
                class="c2"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M128 192l128 128 128-128z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="th"
          colspan="1"
          role="columnheader"
          style="box-sizing: border-box; flex: 20 0 auto; min-width: 30px; width: 20px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
        >
          <div
            class="sortable-column"
          >
            Status
            <div
              class="c0"
              data-testid="sort-column-statusTypeCode"
            >
              <svg
                class="c1"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="margin-bottom: -8px;"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M128 320l128-128 128 128z"
                />
              </svg>
              <svg
                class="c2"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M128 192l128 128 128-128z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="th"
          colspan="1"
          role="columnheader"
          style="box-sizing: border-box; flex: 17 0 auto; min-width: 17px; width: 17px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
        >
          <div
            class="sortable-column"
          >
            Actions
          </div>
        </div>
      </div>
    </div>
    <div
      class="tbody"
      role="rowgroup"
    >
      <div
        class="tr-wrapper"
      >
        <div
          class="tr"
          role="row"
          style="display: flex; flex: 1 0 auto; min-width: 177px;"
        >
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 40 0 auto; min-width: 40px; width: 40px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          />
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 40 0 auto; min-width: 30px; width: 40px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          >
            <span
              data-testid="document-view-filename-text"
            />
          </div>
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 20 0 auto; min-width: 30px; width: 20px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          >
            Jul 27, 2022
          </div>
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 20 0 auto; min-width: 30px; width: 20px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          >
            dbo
          </div>
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 20 0 auto; min-width: 30px; width: 20px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          />
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 17 0 auto; min-width: 17px; width: 17px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          >
            <div
              class="c3 no-gutters row"
            />
          </div>
        </div>
      </div>
      <div
        class="tr-wrapper"
      >
        <div
          class="tr"
          role="row"
          style="display: flex; flex: 1 0 auto; min-width: 177px;"
        >
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 40 0 auto; min-width: 40px; width: 40px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          />
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 40 0 auto; min-width: 30px; width: 40px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          >
            <span
              data-testid="document-view-filename-text"
            />
          </div>
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 20 0 auto; min-width: 30px; width: 20px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          >
            Sep 8, 2022
          </div>
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 20 0 auto; min-width: 30px; width: 20px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          >
            admin
          </div>
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 20 0 auto; min-width: 30px; width: 20px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          />
          <div
            class="td"
            role="cell"
            style="box-sizing: border-box; flex: 17 0 auto; min-width: 17px; width: 17px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
            title=""
          >
            <div
              class="c3 no-gutters row"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  .c0 {
  margin-top: 0.3rem;
}

<div
    class="row"
  >
    <div
      class="align-self-center col-auto"
    >
      <span>
        1 - 2 of  2
      </span>
    </div>
    <div
      class="ml-auto align-self-center col-auto"
    />
    <div
      class="align-self-center col-auto"
    >
      <div>
        <div
          class="Menu-root"
        >
          <div
            class="Menu-button"
          >
            <div
              style="display: flex;"
            >
              <span
                class="c0"
              >
                Show
              </span>
              <input
                class="form-control form-control-sm"
                style="width: 50px; margin-left: 10px; margin-right: 10px;"
                type="number"
                value="10"
              />
              <span
                class="c0"
              >
                Entries
              </span>
            </div>
          </div>
          <div
            class="Menu-items"
            style="width: 10.0rem;"
          >
            <div
              class="Menu-options scrollable list-group"
            >
              <div
                class="Menu-item list-group-item"
                title="menu-item-5"
              >
                5
              </div>
              <div
                class="Menu-item list-group-item"
                title="menu-item-10"
              >
                10
              </div>
              <div
                class="Menu-item list-group-item"
                title="menu-item-20"
              >
                20
              </div>
              <div
                class="Menu-item list-group-item"
                title="menu-item-50"
              >
                50
              </div>
              <div
                class="Menu-item list-group-item"
                title="menu-item-100"
              >
                100
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="align-self-center col-auto"
    >
      <ul
        class="pagination"
      >
        <li
          class="page-item disabled"
        >
          <a
            aria-disabled="true"
            aria-label="Previous page"
            class="page-link"
            rel="prev"
            role="button"
            tabindex="0"
          >
            &lt;
          </a>
        </li>
        <li
          class="page-item active"
        >
          <a
            aria-current="page"
            aria-label="Page 1 is your current page"
            class="page-link"
            role="button"
            tabindex="0"
          >
            1
          </a>
        </li>
        <li
          class="page-item disabled"
        >
          <a
            aria-disabled="true"
            aria-label="Next page"
            class="page-link"
            rel="next"
            role="button"
            tabindex="0"
          >
            &gt;
          </a>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
